src/Transform/ONNX/ConstProp.cpp

```diff
@@ -38,508 +40,739 @@ namespace {
 // There is currently support for adding constant propagation for unary and
 // binary athythmetic ops (binary ops support broadcast). To add an operation,
 // you simply have to add a templated method on how to compute the result in
-// terms of one or two inputs. Values comes as Attribtues, and return is also an
-// Attribute. In that function, presumably you will need different methods to
-// handle int / float / strings... Note that these methods cannot fail. It is
-// your responsablitity to tests for which data type are supported in the rules
-// directly. Specific type restrictions can be added in the DRR files.
-
+// terms of one or two inputs.
+//
 // The methods are:
 //
-// ComputeConstPropElementwiseBinary and ComputeConstPropElementwiseUnary
-// and they need to be tempalted wtih an ONNX Operation (presuably).
+// ElementWiseBinaryOpImpl and ElementWiseUnaryOpImpl
+// and they need to be templated with an ONNX Operation (presuably).
 //
 // Then you need to add rules on how to transform the patterns; look into
 // ConstProp.td for example.
 //
 
-/// A helper function to contruct a RankedTensorType from a ShapedType.
-RankedTensorType constructRankedTensorType(ShapedType type) {
-  assert(type.hasRank() && "Not a ranked type");
-  return RankedTensorType::get(type.getShape(), type.getElementType());
-}
+const StringRef FILE_NAME_ATTR = "file_name";
```

Comment:
@tungld Can you comment on how saving in a separate file works with onnx-mlir-opt when we go from one phase to the next? Is that handled automatically, or do we need to do something separate about the file?

Reply:
It is handled automatically and done inside a single phase of `constprop-onnx`. In a single phase of `constprop-onnx`, we create a file at the first time we reach a constant, and remove this file and return a constant with a dense attribute at the end of the phase. Users of onnx-mlir-opt do not see this file, but a normal constant op.

Reply:
I am not familiar with consprop-onnx: is that one phase of onnx-mlir? So you are saying that each time that constant-prop is called (and it may be called many times), a file is created, store temporary info in it, and then at the end we pluck from that file the constant that are still live, and put them back into mlir as dense attributes? 
In the old approach, you reported mem footprints of several GBs; is that the case now for this file? Or because it is used only for one phase of constant-prop, it is released between such phases and thus is smaller in general.

Reply:
[`constprop-onnx`](https://github.com/onnx/onnx-mlir/blob/master/src/InitOMPasses.hpp#L26) is a pass in onnx-mlir. Creating a file attribute and removing it happen **inside** a single call of `constprop-onnx` pass only, it does not propagate to next calls.

>In the old approach, you reported mem footprints of several GBs; is that the case now for this file? Or because it is used only for one phase of constant-prop, it is released between such phases and thus is smaller in general.

Both approaches do not deal with sharing information between multiple calls of `constprop-onnx`, and `bin/onnx-mlir` command currently calls `constprop-onnx` once. 
The old approach is only for eliminating attribute creation when reading attribute values from constants, but dense attributes are still created for intermediate result constants, while this new approach uses external files for the intermediate result constants.